### PR TITLE
win_firewall: check-mode support, integration tests

### PR DESCRIPTION
**Pull request by dagwieers**
##### SUMMARY
This PR includes:
- Check-mode implementation
- Documentation improvements
- Ensure module output is consistent (no matter what profiles are
  provided)
- Fixed indentation
- Cosmetic changes
- Integration tests

##### ISSUE TYPE

 - Feature Pull Request

##### COMPONENT NAME

win_firewall

##### ANSIBLE VERSION
v2.4

### CODE REVIEW

**Comment by jborean93** [comment]
LGTM
shipit

**Comment by mikeeaton83** [comment]
shipit

File: lib/ansible/modules/windows/win_firewall.ps1 [hunk 1]
```diff
@@ -19,50 +19,51 @@
...
+$ErrorActionPreference = "Stop"
 
-# get params
-$params = Parse-Args $args -supports_check_mode $false
+$params = Parse-Args $args -supports_check_mode $true
+$check_mode = Get-AnsibleParam -obj $params -name "_ansible_check_mode" -type "bool" -default $false
 
-$profiles = Get-AnsibleParam -obj $params -name "profiles" -type "list" -default [ "Public", "Domain", "Private" ]
-$wantedstate = Get-AnsibleParam -obj $params -name "state" -type "str" -failifempty $true -validateset 'enabled', 'disabled'
+$state = Get-AnsibleParam -obj $params -name "state" -type "str" -failifempty $true -validateset 'disabled','enabled'
```
Comment (jhawkesworth) [inline lib/ansible/modules/windows/win_firewall.ps1]: $wantedstate seems to have been changed to $state here (I prefer $state too), but $wantedstate still used on L47  Seems likely to break things.

File: lib/ansible/modules/windows/win_firewall.ps1 [hunk 2]
```diff
@@ -19,50 +19,51 @@
...
 }
 
 Try {
 
-  ForEach($profile in $profiles)
+  ForEach ($profile in @('Domain', 'Private', 'Public')) {
+
+    $currentstate = (Get-NetFirewallProfile -Name $profile).Enabled
+    $result.$profile = @{
+        enabled = ($currentstate -eq $true)
```
Comment (jhawkesworth) [inline lib/ansible/modules/windows/win_firewall.ps1]: would 
enabled = $currentstate
work here?  it would be easier to read and understand if so.

**Comment by jhawkesworth** [comment]
Would be great to have check mode for this module, but I think the $state / $wantedstate thing looks like it is unintended at best.  @dagwieers can you take a look?

File: lib/ansible/modules/windows/win_firewall.ps1 (hunk 1 above)
Comment (dagwieers) [reply lib/ansible/modules/windows/win_firewall.ps1]: Good catch. I still intend to add integration tests to cover this before it is acceptable.

File: lib/ansible/modules/windows/win_firewall.ps1 (hunk 2 above)
Comment (dagwieers) [reply lib/ansible/modules/windows/win_firewall.ps1]: I thought so too, but $currentstate translates to 0 or 1, and I really preferred to have a boolean.

**Comment by dagwieers** [comment]
Integration tests are still missing. I think we should deny new modules that do not ship integration tests (if they don't need any special infrastructure, that is).

**Comment by jborean93** [comment]
@dagwieers I agree, I'm starting to go through all the modules and add in the integration tests (hopefully you've seen the PRs). Hopefully by the time 2.4 comes around we will at least have tests for all modules that don't require special infra.

**Comment by dagwieers** [comment]
@jborean93 Great, I wanted to start on that as well. Please mark the ones you are working on inside the EtherPad page, so we can track who is working on what.

At the rate we are doing stuff these days, there's nothing to be discussed in London ;-)

**Comment by jborean93** [comment]
There's always stuff to be done, have updated the EtherPad with what I've done so far and a link to the PR with the tests.

**Comment by dagwieers** [comment]
@if-meaton @jborean93 @jhawkesworth @nitzmahone Ready to be re-reviewed and merged.

It now includes integration tests, which revealed that we need WMF5+ support for this module to work.
So the integration test have been tested on some Windows 10 systems only.

**Comment by mikeeaton83** [comment]
looks good

shipit

**Comment by dagwieers** [comment]
shipit

**Comment by dagwieers** [comment]
bot_status

**Comment by mikeeaton83** [comment]
shipit

**Comment by dagwieers** [comment]
bot_status

**Comment by jhawkesworth** [comment]
@dagwieers @if-meaton In the past, I think modules which didn't work with WMF 3.0 / powershell 3 have been declined, or rather modified until they worked with everything from 'WMF 3.0 / powershell 3'.  If I recall, last time this was discussed it was felt  WMF 3.0 / powershell 3 would continue to be the lowest version until the versions of windows they run on are no longer supported by MS (some time in 2020 I think, from memory).

Its not the purpose of this PR to change the module, though.

I think this should probably be merged as is, but we should at the very least document that WMF 5 is required (and preferably make it work with older windows if possible).  Its worth doing the documentation change as it will head off some bug reports once the module arrives in a released version of ansible.  

Better still would be to do the get feature check in the module code and have it fail-json with a message saying WMF 5.0 is needed for this module - that would likely mean changing the tests as well of course.

**Comment by dagwieers** [comment]
@jhawkesworth This PR already includes a note saying that WMF 5+ is needed. But I agree, if we can make it work for something older, we should. The cmdlet however does not seem to exist on our Windows 2008 targets.

**Comment by dagwieers** [comment]
@nitzmahone @jborean93 @jhawkesworth @trondhindenes Ready to be reviewed and merged.

**Comment by jhawkesworth** [comment]
yep, there's probably some old way of doing the same thing via WMI.
I'm gonna merge this one as is and document the wmf 5 requirement in a separate PR.

**Comment by dagwieers** [comment]
@jhawkesworth ~I think WMF5 is already documented. I did what you asked.~ Correction, I only add the error, I did not update the python documentation :-(

**Comment by jhawkesworth** [comment]
No worries, I made a PR for that https://github.com/ansible/ansible/pull/25676

**Comment by rruben76** [comment]
I would like to know why WMF5 is required for this Module as the cmdlets Get-NetFirewallProfile and Set-NetFirewallProfile are available as of Windows 2012 with WMF 3.0. Wouldn´t it be possible to modify the If-statement to check for a minimum Windows version (6.2) this would enable this module for Windows 2012 and forward. Or is there another reason why this module is only enabled for WMF5?

**Comment by dagwieers** [comment]
@rruben76 Feel free to improve the module so it works on WMF3 and newer too.

According to https://github.com/ansible/ansible/pull/25127#issuecomment-306451146 it appeared that the integration tests failed for anything older than WMF5, but possibly that is fixable ? I can't remember.

**Pull request merged**